### PR TITLE
Add OGRGeometry::hasEmptyParts()/removeEmptyParts(), and use it in exportToGEOS()

### LINKED ISSUE #10217: `Centroid` operation crashes or hangs
### What is the bug?

Calling `Centroid` on a certain geometry object hangs on Linux or crashes with a segmentation fault on Mac.

### Steps to reproduce the issue

```
(.venv) enf@Erica-Felt pipeline % gdal_translate --version
GDAL 3.9.0, released 2024/05/07
(.venv) enf@Erica-Felt pipeline % python3
Python 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 15.0.0 (clang-1500.1.0.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import osgeo.ogr
>>> geom = osgeo.ogr.CreateGeometryFromWkb(bytearray.fromhex('0106000000010000000103000000020000002100000043ae774ebb642441db70110263994440d7fc784ebb64244184e3675c64994440a1217b4ebb642441741908ab6599444010487e4ebb6424415b5ce0b066994440aa44824ebb6424414ee49593679944405098844ebb6424411762b40868994440080a874ebb6424415c3569946899444088de8a4ebb64244189fc113f699944408d9b8e4ebb642441b73869c2699944409c10934ebb6424414b15451f6a9944403c12984ebb6424414785c0456a994440d895cb4ebb64244188820487699944403cc3494fbb642441317718bf679944402c31784fbb642441607bff4e679944403f547d4fbb64244146d3ed11679944401392824fbb6424417928f2bf669944400a8d874fbb6424416657473e66994440422a8c4fbb6424417ce7549f6599444088d18e4fbb64244134f0ce1765994440cc0d914fbb64244128d86d8a64994440a5ac944fbb642441ea360e7d63994440b96e974fbb642441d4bfab4162994440823c994fbb642441a88b10f8609944402e39994fbb64244182a09a945899444028dd974fbb642441fbf90a4b5399444044f6964fbb64244125c8a7b8519944407895844fbb64244194b3612829994440c5197a4fbb6424418c309ca221994440aa61554ebb642441c1d9475c259944409d665a4ebb64244131f841b829994440f2496d4ebb642441b9ae5aac3a99444046a9724ebb642441fb7937ee4e99444043ae774ebb642441db7011026399444000000000'))
>>> geom.Centroid()
zsh: segmentation fault  python3
```

### Versions and provenance

GDAL 3.9.0, released 2024/05/07 (via homebrew)
```
Darwin Erica-Felt.local 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:19 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8103 arm64
```
MacOS Ventura 13.4.1 (c)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rouault**
Fixes #10217

### CODE REVIEW

File: ogr/ogr_geometry.h [hunk 1]
```diff
@@ -608,6 +608,9 @@ class CPL_DLL OGRGeometry
 
     OGRGeometry *SetPrecision(double dfGridSize, int nFlags) const;
 
+    virtual bool hasEmptyParts() const;
```
Comment (elpaso) [inline ogr/ogr_geometry.h:611]: It seems to me that there is a mix of styles regarding the name of methods: should they start with a capital letter or not?

**Review by elpaso — APPROVED** [review]

File: ogr/ogr_geometry.h (hunk 1 above)
Comment (rouault) [reply ogr/ogr_geometry.h:611]: Yes, I had the same thought and can't really find the nominal pattern here... In most of the rest of the code base methods generally start with capital letter, but OGRGeometry has always had that mix

**Comment by dbaston** [comment]
Does this have applications other than #10217 ? I could certainly backport the GEOS fix to supported branches.

**Comment by rouault** [comment]
> Does this have applications other than #10217

this was the main motivation, but there might be some uses in "cleaning" such weird geometries

**Comment by dbaston** [comment]
I've backported the GEOS fix to 3.9. Unfortunately I just missed the point releases for 3.11 and 3.12.

I have some hesitation about always modifying the geometry before it is sent to GEOS, but I can't point to any actual problem with it.

**Comment by rouault** [comment]
> I have some hesitation about always modifying the geometry before it is sent to GEOS

yes I too. I can restrict doing that to the Centroid() case (if people run GDAL against a GEOS that isn't going to be updated)

**Comment by rouault** [comment]
> yes I too. I can restrict doing that to the Centroid() case

done

**Pull request merged**